### ISSUE TITLE
代码以及CVLoss 的问题

### ISSUE THREAD

**Issue by violet2422** [alexhe101/FAME-Net#10]
https://github.com/alexhe101/FAME-Net/blob/4bbcb551e1a34b55e0c878a749379a19ac115d26/model/moe.py#L271-L283
这里返回的cof_k 的 shape[1] =1 
https://github.com/alexhe101/FAME-Net/blob/4bbcb551e1a34b55e0c878a749379a19ac115d26/solver/moesolver.py#L87-L96
在计算CVLoss 的使用 求取 torch.std 出来的值肯定是nan 这里的代码是不是有问题
下面有示例

![Image](https://github.com/user-attachments/assets/80e179a7-e35f-4fe2-a08b-92c23f124880)

![Image](https://github.com/user-attachments/assets/726321b9-5644-4590-9462-6d681b7ca22c)

**Comment by violet2422** [comment]
https://github.com/alexhe101/FAME-Net/blob/4bbcb551e1a34b55e0c878a749379a19ac115d26/model/moe.py#L248-L260
类HfExpert/LfExpert的 forward 方法中也是一样的问题，

**Comment by mengyuanping** [comment]
> [FAME-Net/model/moe.py](https://github.com/alexhe101/FAME-Net/blob/4bbcb551e1a34b55e0c878a749379a19ac115d26/model/moe.py#L271-L283)
> 
> Lines 271 to 283 in [4bbcb55](/alexhe101/FAME-Net/commit/4bbcb551e1a34b55e0c878a749379a19ac115d26)
> 
>  def forward(self,x): 
>      x = self.pre_fuse(x) 
>      cof = self.gate(x) 
>      out = torch.zeros_like(x).to(x.device) 
>      for idx in range(self.num_experts): 
>          if cof[:,idx].all()==0: 
>              continue 
>          mask = torch.where(cof[:,idx]>0)[0] 
>          expert_layer = self.expert_networks_d[idx] 
>          expert_out = expert_layer(x[mask]) 
>          cof_k = cof[mask,idx].view(-1,1,1,1) 
>          out[mask]+=expert_out*cof_k 
>      return out,cof_k 
> 
> 这里返回的cof_k 的 shape[1] =1
> [FAME-Net/solver/moesolver.py](https://github.com/alexhe101/FAME-Net/blob/4bbcb551e1a34b55e0c878a749379a19ac115d26/solver/moesolver.py#L87-L96)
> 
> Lines 87 to 96 in [4bbcb55](/alexhe101/FAME-Net/commit/4bbcb551e1a34b55e0c878a749379a19ac115d26)
> 
>  class CVLoss(nn.Module): 
>      def __init__(self, loss_weight=1.0,reduction='mean'): 
>          super(CVLoss, self).__init__() 
>          self.loss_weight = loss_weight 
>          self.reduction = reduction 
>      def forward(self,logits): 
>          # print(torch.mean(logits,dim=1).shape) 
>          cv = torch.std(logits,dim=1)/torch.mean(logits,dim=1) 
>          # print(cv) 
>          return self.loss_weight*torch.mean(cv**2) 
> 
> 在计算CVLoss 的使用 求取 torch.std 出来的值肯定是nan 这里的代码是不是有问题
> 下面有示例
> ![Image](https://github.com/user-attachments/assets/80e179a7-e35f-4fe2-a08b-92c23f124880)
> 
> ![Image](https://github.com/user-attachments/assets/726321b9-5644-4590-9462-6d681b7ca22c)

请问您解决了吗？

**Comment by violet2422** [comment]
他这就是错的
&nbsp;


王同学
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2025年10月13日(星期一) 晚上10:09
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [alexhe101/FAME-Net] 代码以及CVLoss 的问题 (Issue #10)












mengyuanping left a comment (alexhe101/FAME-Net#10)
  
FAME-Net/model/moe.py
 
Lines 271 to 283 in 4bbcb55
 
def forward(self,x):
 x = self.pre_fuse(x)
 cof = self.gate(x)
 out = torch.zeros_like(x).to(x.device)
 for idx in range(self.num_experts):
 if cof[:,idx].all()==0:
 continue
 mask = torch.where(cof[:,idx]&gt;0)[0]
 expert_layer = self.expert_networks_d[idx]
 expert_out = expert_layer(x[mask])
 cof_k = cof[mask,idx].view(-1,1,1,1)
 out[mask]+=expert_out*cof_k
 return out,cof_k
 
这里返回的cof_k 的 shape[1] =1
 FAME-Net/solver/moesolver.py
 
Lines 87 to 96 in 4bbcb55
 
class CVLoss(nn.Module):
 def init(self, loss_weight=1.0,reduction='mean'):
 super(CVLoss, self).init()
 self.loss_weight = loss_weight
 self.reduction = reduction
 def forward(self,logits):
 # print(torch.mean(logits,dim=1).shape)
 cv = torch.std(logits,dim=1)/torch.mean(logits,dim=1)
 # print(cv)
 return self.loss_weight*torch.mean(cv**2)
 
在计算CVLoss 的使用 求取 torch.std 出来的值肯定是nan 这里的代码是不是有问题
 下面有示例
 image.png (view on web)
 
image.png (view on web)
  
请问您解决了吗？
